### DISCUSSION TITLE
Syntax highlighting for classes doesn't use the Configurator's "Class" style

### DISCUSSION THREAD

**Discussion by M-AsInMancy** [blu3mania/npp-papyrus#41] [Q&A] [answered]
Been a longtime user of this plugin. It's tremendously helpful and I really appreciate it. 

I just upgraded to v0.5 from v0.4. After updating some of my syntax highlighting appears broken. Specifically classes and forms no longer have unique coloring. The rest of syntax highlighting works like it did before. Allow me to convey the issue with some examples.

`Event OnSomeEvent(String EventName, Float Number, Bool IsValid, Form Sender)`

In NP++ the `Form` portion of that code is colorized like the parameter names. It has the same color as `Sender`. While Float, String, and Bool use the "Type" coloring specified in the Style Configurator. But `Form Sender` uses the Default coloring from the Configurator. And I have confirmed its using the default by adjusting that color and watching it apply to `Form`.

Another example would be a simple line like `Utility.wait(2.0)`. In the previous version Utility would use the Class style for the Configurator. But now it uses the Default style for syntax highlighting. 

I am uncertain what is causing this. I have tried updating NPP. Reinstalling this plugin. Played around with every setting I can think of. 

One other thing is broken. Though it has been for many months. Much longer than the current problem. When I first installed the Lexer the class names would be highlighted as links. And of course I could Ctrl + Double click and they would take me to the class definition. That stopped working. They were no longer highlighted and could not be clicked on. Even though styling class names is enabled in the Lexer Settings. Also I have the class cache disabled. I can not precisely recall when this stopped working. Perhaps around the time I reinstalled Skyrim to a different hard drive. I mostly just ignored this issue. But I think it might be related to my new trouble. I wonder if the new v0.5 is relying on being able to find and link classes in order for the syntax highlighting to function as expected. And since it can't hyperlink classes the style configuration is now broken too. 

How does the Lexer know where to search for class definitions? Does it use the Import directories specified in the compiler settings? Or does it possibly search the default Skyrim installation directory? And fail to find them if you have installed to an alternate location? Thanks for any help! And thanks for a great tool.

**Comment by blu3mania** [comment, marked as answer]
Thanks for the report, and It's indeed a bug introduced in v0.5. I just fixed it and released v0.5.1. I forgot to test it when class name caching is turned off, since the main focus of v0.5 was dark mode.

Yes, the lexer searches for class names in all import directories, so please make sure the configured import directories match your current installation, and the referenced script (e.g. Form.psc) is in one of the import directories.

**Reply by M-AsInMancy** [reply]
Confirmed fix. Thank you! Looking back it is funny that the one thing I did not try was enabling caching. My thinking was that if it is broken without a cache then the cache would store a broken result, and make debugging more difficult. 

Before I go I wanted to pass on my genuine appreciation for this plugin. It has helped write many a mod. I am accustomed to IDEs. But thanks to this plugin and its predecessor, I have felt at home spending countless hours writing Papyrus code. Ta!

**Reply by M-AsInMancy** [reply]
Oh one piece of feedback regarding import directories. It might not hurt to mention the order of precedence somewhere. The directories that are higher in the list override the directories underneath. This is the opposite behavior of the compile command -i. Where your list of directories is separated by a semicolon. In that the first directory gets overridden by the next. So it was my instinct to use the same order of precedence here. But through a little trial and compilation error I determined my mistake. I know space in the UI is limited. Perhaps this is best mentioned in the Configuration.md?

**Reply by M-AsInMancy** [reply]
Guess I am not going away. I found one more minor issue in v0.5.1. When I follow a link to open up a class, the syntax highlighting for the previous document stops working. It begins exhibiting the same symptoms as the original issue. No links and different coloring. If I try switching to a different document tab and then come back to the original a second time, it resumes working again. Very minor issue. 

Steps to reproduce:
1. With cache disabled Ctrl + Double Click on a class link. This opens a new tab.
2. Return to the original tab. Syntax highlighting is broken.

Steps to mitigate after the above:
1. Click on any other document tab to switch away.
2. Click back to the original tab. Syntax highlighting resumes working.

**Reply by blu3mania** [reply]
> Oh one piece of feedback regarding import directories. It might not hurt to mention the order of precedence somewhere. The directories that are higher in the list override the directories underneath. This is the opposite behavior of the compile command -i. Where your list of directories is separated by a semicolon. In that the first directory gets overridden by the next. So it was my instinct to use the same order of precedence here. But through a little trial and compilation error I determined my mistake. I know space in the UI is limited. Perhaps this is best mentioned in the Configuration.md?

Oh, I didn't know the behavior of "-i" is like that, since I have never had conflicting scripts in multiple folder. I will do some testing later to confirm, and if it's the case, I will update the code to match the behavior.

**Reply by blu3mania** [reply]
> Guess I am not going away. I found one more minor issue in v0.5.1. When I follow a link to open up a class, the syntax highlighting for the previous document stops working. It begins exhibiting the same symptoms as the original issue. No links and different coloring. If I try switching to a different document tab and then come back to the original a second time, it resumes working again. Very minor issue.
> 
> Steps to reproduce:
> 
> 1. With cache disabled Ctrl + Double Click on a class link. This opens a new tab.
> 2. Return to the original tab. Syntax highlighting is broken.
> 
> Steps to mitigate after the above:
> 
> 1. Click on any other document tab to switch away.
> 2. Click back to the original tab. Syntax highlighting resumes working.

Hmm, I couldn't reproduce this issue, with the option on or off. I tried small scripts and large scripts, and also tried switching back to original tab or closing the new tab to return to original tab,. In all scenarios the syntax highlighting in the original tab still works fine. The only thing to note is that double clicking selects the text in the original tab (Notepad++ behavior) so when you switch back, all instances of the selected text are highlighted, which can be easily dismissed by clicking on somewhere else. Any other information to help me reproduce?

**Reply by M-AsInMancy** [reply]
@blu3mania after experimenting some more this issue is oddly intermittent. It also happens with or without caching enabled. Anytime I disable or enable the cache I also close the script in question and exit Notepad++. Just to try and limit the possibility that something in memory is the culprit. 

I have noticed the issue is more likely to occur on larger scripts. And it appears more likely to occur the more Papyrus tabs I have open.
In fact sometimes the same script will work fine. Clicking a link does not dehighlight classes. But if I click enough links spawning a whole bunch of tabs now that same script will stop highlighting. 

I also noticed an issue when opening multiple especially large scripts. Like the CW (Civil War) scripts included in the base game. If open some of those while having other large scripts open, the links do not appear nor does the syntax highlighting for them. No matter how many times I leave the tab and come back. This happens with or without cache enabled. If I exit Notepad with these open. Then reopen Notepad (which automatically loads all of my existing tabs) then the link highlights work for the reopened tab! But as soon as I switch tabs the links stop working. This is a slightly different issue than before. And it happens with and without the cache enabled.

I wish I could be more help in isolating this. Another possible culprit is that I have over 20k scripts in my scripts directory. If this is running out of memory or some other stress test related problem, perhaps it is due to the high amount of classes being processed? 

If you can think of something more for me to try just name it. IMO this is a fairly minor issue. With reasonably sized scripts and only a few tabs open it works fine.

**Reply by M-AsInMancy** [reply]
> Oh, I didn't know the behavior of "-i" is like that, since I have never had conflicting scripts in multiple folder. I will do some testing later to confirm, and if it's the case, I will update the code to match the behavior.

One thing to consider is that if you reverse the order then everyone updating will now have to redo their settings. Just my opinion, your current order is fine. But a little tooltip popup or some additional instruction to indicate the precedence would assist people who are configuring it for the first time. Just a minor enhancement idea.

**Reply by blu3mania** [reply]
I did lots of testing but still couldn't reproduce. I opened 10+ large scripts including CWScript and some others that are even bigger, plus about 30 other scripts, and kept opening other scripts from CWScript. Everything just worked fine. Although I don't have 20k scrips on disk, but I still have about 12k so it's still a lot of files. I guess having 32G memory may have helped? Not sure.

The thing is, if it's really memory related then I can do nothing. It is really controlled by Notepad++ and its Scintilla/Lexilla components. The only dynamic memory accumulation by this plugin is class name caching, which is still a tiny amount (even 20k class names with an average of 10 chars per script is just 200k, not even 1M...), and you already turned it off.

Are you using 32-bit Notepad++ or 64-bit? One thing I can suggest to not having tons of files open at the same time is using plugins such as Session Manager. You can easily have different sessions for different "projects", which also makes it easier to manage when working on multiple MODs.

 As for the searching order of scripts in imported directories, I also did many tests, and I confirmed that PapyrusCompiler actually uses the same searching order (i.e. from first to last). You can easily test it by creating a test script A that extends another test script B and having 2 versions of B, one works and the other one doesn't, placing the 2 versions in different folders, and specifying these folders in "-i" param with different orders. One thing though, PapyrusCompiler searches in current folder before it searches in provided import directories, and this plugin doesn't do that. It may be funny but the way Notepad++ is designed, a lexer plugin doesn't know the exact file it is processing, and it's not an easy task to find out. I have some idea but it may not always be reliable. I will try to add this logic.

**Reply by M-AsInMancy** [reply]
@blu3mania I am using the 64bit edition. Appreciate the tip about the Session Manager. That could be perfect for certain needs.

I will keep an eye on the intermittent behavior and see if I can ever figure out steps to reproduce. As I said before it's so minor it is easy to live with. Now that I have had some time to work in the editor more I have noticed that it will often go away once I type a few new lines of code. Which makes it even easier to live with.

I did not realize that about the compiler. You nailed the cause of my misunderstanding. I always inserted the current dir into the -i command in addition to an external dir containing all other scripts. I guess that was unneeded!

Thanks for the investigation, help, and latest patch.
